### DISCUSSION TITLE
Finding a confusing order of phases in equilibrium calculations

### DISCUSSION THREAD

**Discussion by MWW726** [pycalphad/pycalphad#433] [Q&A] [answered]
I am doing an work about numerical calculation, and found a confusing order of phases in equilibrium calculations. 

my code is below:
dbf = Database('....../al-cu.TDB')
comps = ['AL', 'Cu', 'VA']  
phases = dbf.phases.keys()
eq = equilibrium(dbf, comps, phases, {v.X('CU'):(0.096), v.T:(874, 875.6, 0.1), v.P:101325}, output='HM')
eq = eq.squeeze()
eq.Phase

result is below:
array([['FCC_A1', 'LIQUID', ''],
       ['FCC_A1', 'LIQUID', ''],
       ['FCC_A1', 'LIQUID', ''],
       ['FCC_A1', 'LIQUID', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', ''],
       ['LIQUID', 'FCC_A1', '']], dtype='<U14')

We can see that the phase order has changed, so is there any solution to this? Otherwise it is a potential trouble. The TDB is also uploaded here. 

[al-cu.txt](https://github.com/pycalphad/pycalphad/files/9859719/al-cu.txt)

**Comment by bocklund** [comment, marked as answer]
Thanks for your question!

The phases are not guaranteed to be in any particular order. There are a few different ways to extract data for one phase at a time. This notebook shows one way https://github.com/pycalphad/cookbook/blob/master/ScreeningPhaseStability/ScreeningPhaseStability.ipynb

If you let us know what you’re trying to do, we can help you use the right approach if the screening phase stability example isn’t quite right for you 😀

**Reply by MWW726** [reply]
Briefly, For a two-phase region(a solid phase and a liquid phase), what I need is eq.X (eq is result of equilibrium(...) function) and I am using eq.X.isel(vertex=n) as the concentration of n-th phase to do some calculations. But, phase order in the eq.Phase could be changed as we talked above, which makes eq.X.isel(vertex=n) is not just the component of a phase. 

I have fixed the problem using the code below(C is component):
phases = ['LIQUID','FCC_A1']
C_phases = list()
for each in phases:
    index = np.argwhere(eq.Phase.values==each)[:,1]
    C_phase = np.empty(shape=(0,N_elements))
    for i in range(len(eq.T)):
        C_phase = np.vstack((C_phase, 
                                       eq.X.sel(vertex=index[i],T=eq.T[i])))
    C_phases.append(C_phase)
da_C =xr.DataArray(np.array(C_phases), coords=[[0,1],eq.T.values, [0,1]], dims=['vertex', 'T','component'])

Now, the DataArray, da_C.isel(vertex=n), is guaranteed to be the component of a phase. But it is better if equilibrium() function could be like this.

**Reply by bocklund** [reply]
It sounds like you’d benefit from the screening phase stability notebook I linked. There’s some examples in there that would help you select only the data for one particular phase in two phase regions by applying masks. 

I didn’t test this snippet, but something like

```python
mask = ((eq.vertex =! “”).sum() == 2) & (eq.vertex == “LIQUID”) 
eq.where(mask).X
```

should create a mask for only data for the liquid phase inside two phase regions, which you can then apply via `where` and access `eq` like usual otherwise.

**Reply by MWW726** [reply]
OK, the mask operation inspired me
